src/app/services/xml-parsers/structure-xml-parser.service.ts

```diff
@@ -21,63 +21,129 @@ export class StructureXmlParserService {
-            }
-
-            return false;
-          });
-
-          const page: Page = {
-            id: element.getAttribute('xml:id') || 'page_' + (pagesIndexes.length + 1),
-            label: element.getAttribute('n') || 'Page ' + (pagesIndexes.length + 1),
-            originalContent: pageContent,
-            parsedContent: pageContent.map(child => this.genericParserService.parse(child as XMLElement)),
-          };
-          pages[page.id] = page;
-          pagesIndexes.push(page.id);
-        }
+      const pagesNumber = pageElements.length;
+
+      if (pagesNumber > 0) {
+        this.parseDocumentPages(pageElements, pagesNumber, pages, pagesIndexes);
       } else {
-        // No <pb> used => TODO: Decide how to handle text division
-        console.warn('TODO: Decide how to handle text division when there are no <pb>s');
-        const mainText = document.querySelector('text');
-        const content = Array.from(mainText.childNodes);
-        const page: Page = {
-          id: `page_${new Date().getTime()}`,
-          label: 'Main Text',
-          originalContent: content as XMLElement[],
-          parsedContent: content.map(child => this.genericParserService.parse(child as XMLElement))
-            // tslint:disable-next-line: no-string-literal
-            .filter(c => !!c['content']), // TODO: FIXME: fix property access
-        };
-        pages[page.id] = page;
-        pagesIndexes.push(page.id);
+        this.parseDocument(document, pages, pagesIndexes);
       }
-      console.log(pages);
     }
 
     return {
       pages,
       pagesIndexes,
     };
   }
+
+  parseDocumentPages(pageElements: NodeListOf<Element>, pagesNumber: number, pages: Map<Page>, pagesIndexes: string[]) {
+    pageElements.forEach((element, i) => {
+      let pageContent: XMLElement[] = [];
+
+      if (pagesNumber === 1) {
+        pageContent = getElementsAfterTreeNode(element);
+      }
+
+      if (pagesNumber !== 1) {
+        const isFirstPage = i === 0;
+        const isLastPage = i === pagesNumber - 1;
+
+        if (isFirstPage) {
+          pageContent = getElementsBetweenTreeNode(element, pageElements[i + 1]);
+        }
+
+        if (isLastPage) {
+          pageContent = getElementsAfterTreeNode(element);
+        }
+
+        if (!isFirstPage && !isLastPage) {
+          pageContent = getElementsBetweenTreeNode(element, pageElements[i + 1]);
+        }
+      }
+
+      pageContent = this.removeBackNodes(pageContent);
+
+      const page: Page = {
+        id: element.getAttribute('xml:id') || 'page_' + (pagesIndexes.length + 1),
+        label: element.getAttribute('n') || 'Page ' + (pagesIndexes.length + 1),
+        originalContent: pageContent,
+        parsedContent: this.parsePageContent(pageContent),
+      };
+      pages[page.id] = page;
+      pagesIndexes.push(page.id);
+    });
+  }
+
+  parseDocument(document: XMLElement, pages: Map<Page>, pagesIndexes: string[]) {
```

Comment:
Same as the previous comment, I would expected to have a `parseDocumentAsPage` method that given an XMLElement returns a `Page` element, so that it could be used also for single XMLElements without affecting any parsed data collector variable.